cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -357,6 +358,9 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
           systemArgs.put(ProgramOptionConstants.CLUSTER_KEY_INFO,
                          properties.get(ProgramOptionConstants.CLUSTER_KEY_INFO));
         }
+        systemArgs.put(ProgramOptionConstants.ARTIFACT_ID,
```

Comment:
if these things are needed in the system args, they should be placed there at the start of the run (the runInternal method), not later on in the lifecycle. This probably goes for whatever else is being updated within the AbstractProgramRuntimeService. Is there a reason they can't be set at the very beginning?